client/castApp/src/castinit.c

```diff
@@ -65,13 +67,63 @@ static void casthook(initHookState state)
     epicsAtExit(&castexit, NULL);
 }
 
+/*
+    Example call: addReccasterEnvVars("SECTOR") or addReccasterEnvVars("SECTOR", "BUILDING")
+    Appends the given env variables to the extra_envs list to be sent in addition to the default_envs array
+*/
+static void addReccasterEnvVars(int argc, char **argv)
+{
+    if(argc == 1) {
+        errlogSevPrintf(errlogMinor, "At least one argument expected for %s\n", __func__);
+        return;
+    }
+    int i;
+    /* skip first arg since that is the function name */
+    for(i = 1; i < argc; i++) {
+        if(argv[i] == NULL) {
+            errlogSevPrintf(errlogMinor, "Arg is NULL for %s\n", __func__);
+            continue;
+        }
+        if(argv[i][0] == '\0') {
+            errlogSevPrintf(errlogMinor, "Arg is empty for %s\n", __func__);
+            continue;
+        }
+        epicsMutexMustLock(thecaster.lock);
+        thecaster.extra_envs = realloc(thecaster.extra_envs, sizeof(char *) * (++thecaster.num_extra_envs + 1));
+        epicsMutexUnlock(thecaster.lock);
+
+        if (thecaster.extra_envs == NULL) {
+            errlogSevPrintf(errlogMajor, "Error in memory allocation of extra_envs from %s", __func__);
+            return;
+        }
+        const size_t slen = strlen(argv[i]) + 1;
+        char *newvar = (char *)calloc(slen, sizeof(char));
+        strncpy(newvar, argv[i], slen);
+
+        epicsMutexMustLock(thecaster.lock);
+        thecaster.extra_envs[thecaster.num_extra_envs - 1] = newvar;
+        thecaster.extra_envs[thecaster.num_extra_envs] = NULL;
```

Comment:
yes this is true. Do you think there is any benefit with having the extra safety of being null terminated?

Reply:
Having a `realloc()` in this loop isn't so efficient,  Also, having multiple lock/unlock operations in the loop seems like an unnecessary invitation for bugs.  This requires that code accessing `extra_envs[]` be aware that some elements may be `NULL` (and perhaps not just the last).

Personally, I see "coarse" locking as a safer starting position (aka. lock over a much as possible).  imo. Finer grained locking should be seen as an optimization when necessary, and with due consideration given to possible concurrent access.

Also, wrt. `var = realloc(var, ...);` keep in mind that `realloc()` does __not__ `free()` anything on error.  So using the same variable for both result and argument is an easy way to leak on error.  (recent GCC should warn about this)

If you want to optimize to avoid all of the `calloc()`s while locked, then I would suggest making two loops.  One without locking, which looks only at the argument list.  Then a second with locking, to modify the `caster_t*`.

eg. the following skeleton, which omits printing error messages

```c
static void addReccasterEnvVars(caster_t* self, int argc, char **argv) {
  int i;
  int ret = 0;
  char **newelems;
  if(argc<2)
    return; // one arg required
  /* temporary newelems[] to hold individually allocated strings */
  if(!(newelems = calloc(argc-1, sizeof(*newelems))))
    ret = 1;
  for(i=0; i<argc-1 && !ret; i++) {
    if(!(newelems[i] = strdup(argv[i+1])))
      ret = 1;
  }
  if(!ret) {
    char **new_envs;
    epicsMutexMustLock(self->lock);
    if(self->shutdown) {
      /* shutdown in progress, silent no-op */
    } else if(self->state != casterStateInit) {
      /* Attempt to add after iocInit(), when we may be connected.
       * To fully support, would need to force reconnect or resend w/ updated envs list.
       */
       ret = 2;
    } else if(!(new_envs = realloc(self->extra_envs,
                                   sizeof(*new_envs) * (thecaster.num_extra_envs + argc - 1)))) {
      ret = 1;
    } else {
        /* from this point, nothing can fail */
        self->extra_envs = new_envs;
        thecaster.num_extra_envs += argc - 1;

        for(i=0; i<argc-1; i++) {
          new_envs[thecaster.num_extra_envs + i] = newelems[i];
          newelems[i] = NULL; /* prevent early free below */
        }
    }
    epicsMutexUnlock(self->lock);
  }
  /* cleanup newelems[] on success or failure */
  for(i=0; i<argc-1 && !ret; i++) {
    free(newelems[i]);
  }
  free(newelems);
  if(ret) ... oops
```

Note that this skeleton ignores the possibility of duplicate variable names.

Reply:
Thanks @mdavidsaver I'll have a look more at your example. Just quick question:

> which omits printing error messages

Should I take this as meaning I shouldn't print error messages and just return instead?

Reply:
> Should I take this as meaning I shouldn't print error messages and just return instead?

Not at all!  My "example" is long enough without wordy prints.

Reply:
Hi @mdavidsaver ,

I pushed a commit to try to address this and worked off your skeleton. I had some questions while writing it up:

In your skeleton I see a check on `!ret` for the cleanup of the temporary array. I think we don't want this check since we want to clean up that memory either way?
```
  /* cleanup newelems[] on success or failure */
  for(i=0; i<argc-1 && !ret; i++) {
```

I added logic to check if someone tries to add an environment variable that is already in `self->extra_envs`. This could be extended to also check against the default_envs list too but that isn't in there right now. With this logic check, I had to add a second lock/unlock for accessing self->extra_envs

After adding the second lock, I thought about moving the checks on this->shutdown and this->current to the top and created a commit for that. I didn't end up going with it since I thought there might be good reason to keep those checks in the same lock/unlock block with where we edit `self->extra_envs`. But if you are curious you can view that idea here: https://github.com/tynanford/recsync/commit/f566ba228d54c42b946496e78dcbe79e563075b0

Appreciate the help